src/backend/OpenCL/CHIPBackendOpenCL.cc

```diff
@@ -938,22 +997,36 @@ CHIPEvent *CHIPQueueOpenCL::launchImpl(CHIPExecItem *ExecItem) {
 
   logTrace("Launch LOCAL: {} {} {}", Local[0], Local[1], Local[2]);
 #ifdef DUBIOUS_LOCKS
-  LOCK(Backend->DubiousLockOpenCL)
+  LOCK(Backend->DubiousLockOpenCL);
 #endif
+
+  auto SvmAllocationsToKeepAlive =
+      annotateSvmPointers(*OclContext, Kernel->get()->get());
+
   auto Status = clEnqueueNDRangeKernel(ClQueue_->get(), Kernel->get()->get(),
                                        NumDims, GlobalOffset, Global, Local, 0,
                                        nullptr, LaunchEvent->getNativePtr());
   CHIPERR_CHECK_LOG_AND_THROW(Status, CL_SUCCESS, hipErrorTbd);
 
-  if (std::shared_ptr<CHIPArgSpillBuffer> SpillBuf =
-          ExecItem->getArgSpillBuffer()) {
-    // Use an event call back to prolong the lifetime of the spill buffer
-    // in case the exec item gets destroyed before the kernel
-    // is launched/completed (may happen when called from
-    // CHIPQueue::launchKernel()).
-    auto *CBData = new decltype(SpillBuf)(SpillBuf);
+  std::shared_ptr<CHIPArgSpillBuffer> SpillBuf = ExecItem->getArgSpillBuffer();
+
+  if (SpillBuf || SvmAllocationsToKeepAlive) {
+    // Use an event call back to prolong the lifetimes of the
+    // following objects until the kernel terminates.
+    //
+    // * SpillBuffer holds an allocation referenced by the kernel
+    //   shared by exec item, which might get destroyed before the
+    //   kernel is launched/completed
+    //
+    // * Annotated SVM pointers may need to outlive the kernel
```

Comment:
Isn't it down to the program logic to not free SVM allocations while they might be in use?

Reply:
All SVM allocations from all threads are annotated as we can’t tell which ones are needed for the kernel launch. One thread may free a SVM allocation while another thread still holds it.

Reply:
OK, I see -- annoying. Did you manage to crash an OpenCL implementation by freeing an SVM allocation while/before the kernel was running? Sounds like an undefined corner case indeed.

Reply:
We might want to update the OpenCL spec for this and make it defined behavior: If the SVM alloc is freed during or before, the implementation should ignore it when synchronizing the data.